standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1201,22 +1199,86 @@ private TableValidWriteIds getValidWriteIdsForTable(Statement stmt, String fullT
     }
   }
 
+  private List<Long> getValidTxnIds(AllocateTableWriteIdsRequest rqst, Statement stmt, ResultSet rs)
+          throws MetaException, SQLException {
+    List<Long> txnIds = new ArrayList<>();
+    String dbName = rqst.getDbName().toLowerCase();
+    String tblName = rqst.getTableName().toLowerCase();
+    Long validWriteId;
+    Long largestWriteId, smallestWriteId;
+    List<TxnToWriteId> txnToWriteIds = rqst.getTxnToWriteIdList();
+    assert (txnToWriteIds != null);
+
+    // Get the next write id for the table and check if its in consistent with the input txntowriteids.
+    String s = sqlGenerator.addForUpdateClause(
+            "select nwi_next from NEXT_WRITE_ID where nwi_database = " + quoteString(dbName)
+                    + " and nwi_table = " + quoteString(tblName));
+    LOG.debug("Going to execute query <" + s + ">");
+    rs = stmt.executeQuery(s);
+    if (rs.next()) {
+      validWriteId = rs.getLong(1);
+    } else {
+      // All input write ids should be greater than this
+      validWriteId = 0L;
+    }
+
+    int size = txnToWriteIds.size();
+    for (TxnToWriteId txnToWriteId :  txnToWriteIds) {
+      Long targetTxnId = getTargetTxnId(rqst.getReplPolicy(), txnToWriteId.getTxnId(), stmt);
+      if (targetTxnId == -1) {
+        LOG.info("Transaction " + txnToWriteId.getTxnId() +
+                " is invalid, may be already committed/aborted. Ignoring the whole operation");
+        return new ArrayList<>();
+      }
+
+      Long writeId = txnToWriteId.getWriteId();
+      if (writeId <= validWriteId) {
+        //looks like a retry, so do a noop
+        LOG.info("Write Id " + writeId + " is less than valid write id : " + validWriteId + "Looks like retry.");
+        return new ArrayList<>();
+      }
+
+      if (writeId > (validWriteId + size)) {
+        //this should never happen in case of replication
+        LOG.info("Write Id " + writeId + " is out of range. Valid write id : " + validWriteId + " Size : " + size);
+        throw new MetaException("Write Id " + writeId + " is out of range. Valid write id : " +
```

Comment:
Error message can be clear like "Replaying allocation of write id failed as current writeid: " + writeId " is not replayed in sequence. Probably, some of the allocate write id events are skipped or lost."